crucible-mir: fix Crucible signature of `extern "rust-call"` functions

spernsteiner opened the pull request:
The unstable `extern "rust-call"` ABI used for closures is normally compiled by untupling the last argument: `extern "rust-call" fn(T, (U, V))` is compiled roughly the same as `fn(T, U, V)`.  Some parts of rustc's MIR code generation depend on this equivalence to perform operations that would otherwise be invalid.  This branch replaces our previous workaround (the "`spread_arg` hack") with a new implementation that behaves more like rustc's existing backends.  In particular, with this branch, the tupled and untupled signatures now map to the same Crucible function signature.  This also lets us add support for calling closures through `&dyn Fn`.

With this change, the handling of `extern "rust-call"` functions in crux-mir is more consistent, but there is still one tricky case.  At the MIR level, `extern "rust-call"` functions are normally defined with a tupled signature and called by passing a tuple of arguments.  However, the implementation functions of closures (e.g. for `|x| x + 1`, the MIR function that computes `x + 1`) are defined with an untupled signature, but are still passed a tuple at the call site.  This case is detected by looking for the combination of `"abi": {"kind": "RustCall"}` and `"spread_arg": null` on the function definition, and it's handled separately from `RustCall` with non-null `spread_arg` in some places.

Fixes #1222

RyanGlScott commented on crucible-mir/src/Mir/Mir.hs:
```diff
@@ -133,28 +133,33 @@ data FnSig = FnSig {
+    deriving (Show, Eq, Ord, Generic)
+
+data RustCallBodyInfo
+    = RcNoBody
+    -- ^ The `FnSig` containing this `RustCall` `Abi` isn't the signature of a
+    -- `Fn` body.  For example, `extern "rust-call"` function pointers always
+    -- have an ABI of `RustCall RcNoBody`.
+    | RcNoSpreadArg
+    -- ^ The `FnSig` is associated with a body, but the body's `spread_arg`
+    -- field is unset.  This applies to closure implementations, which are
+    -- are `extern "rust-call"` and are called with tupled arguments, but are
+    -- defined with un-tupled arguments (and no `spread_arg`).  For such
+    -- functions, `fsarg_tys` will contain the un-tupled arguments, and the ABI
+    -- will be `RustCall RcNoSpreadArg`.
+    | RcSpreadArg Int
+    -- ^ The `FnSig` is associated with a body, and the body has a `spread_arg`
+    -- index.  This applies to various `extern "rust-call"` closure-related
+    -- shims and vtable methods, which are both called and defined with a
+    -- tupled signature.
```
Haddock nit pick: surrounding something with backticks causes Haddock to try and hyperlink it as if it were a Haskell identifier. Use `@...@` for general monospace text:

```suggestion
    = RcNoBody
    -- ^ The `FnSig` containing this `RustCall` `Abi` isn't the signature of a
    -- `Fn` body.  For example, @extern \"rust-call\"@ function pointers always
    -- have an ABI of @`RustCall` `RcNoBody`@.
    | RcNoSpreadArg
    -- ^ The `FnSig` is associated with a body, but the body's @spread_arg@
    -- field is unset.  This applies to closure implementations, which are
    -- are @extern \"rust-call\"@ and are called with tupled arguments, but are
    -- defined with un-tupled arguments (and no @spread_arg@).  For such
    -- functions, @fsarg_tys@ will contain the un-tupled arguments, and the ABI
    -- will be `RustCall RcNoSpreadArg`.
    | RcSpreadArg Int
    -- ^ The `FnSig` is associated with a body, and the body has a @spread_arg@
    -- index.  This applies to various @extern \"rust-call\"@ closure-related
    -- shims and vtable methods, which are both called and defined with a
    -- tupled signature.
```

Also, you have to escape double quotes, since Haddock treats `"..."` as though you are trying to hyperlink a Haskell module name.

samcowger replied:
I think the fact that the string is hyphenated will actually stop Haddock from trying to turn it into a link - though the reference is [a bit vague](https://haskell-haddock.readthedocs.io/latest/markup.html#linking-to-modules) on the rules for such links:
> A basic check is done on the syntax of the header name to ensure that it is valid before turning it into a link

spernsteiner replied:
Fixed in 35f00b44baee28f5649bba773df875c95d3b71be

RyanGlScott commented on crucible-mir/src/Mir/JSON.hs:
```diff
@@ -232,10 +230,17 @@ instance FromJSON Collection where
 instance FromJSON Fn where
     parseJSON = withObject "Fn" $ \v -> do
       args <- v .: "args"
+      abi <- v .: "abi" >>= \abi -> case abi of
+        RustCall _ -> do
+          spreadArg <- v .: "spread_arg"
+          case spreadArg of
+            Just i -> return $ RustCall (RcSpreadArg i)
+            Nothing -> return $ RustCall RcNoSpreadArg
+        _ -> return abi
```
There's a lot of infix operators on the `abi <- v .: "abi" >>= \abi -> case abi of` line. How about splitting this up a bit:

```hs
      abi <- v .: "abi"
      let abi' =
            case abi of
              RustCall _ -> do
                spreadArg <- v .: "spread_arg"
                case spreadArg of
                  Just i -> return $ RustCall (RcSpreadArg i)
                  Nothing -> return $ RustCall RcNoSpreadArg
              _ -> return abi
```

spernsteiner replied:
Fixed in 579afd3857198af9fc108870f387483beb58c677

RyanGlScott commented on crucible-mir/src/Mir/PP.hs:
```diff
@@ -135,12 +135,15 @@ pretty_temp (Var vn vm vty _vzst) =
         rty    = fs^.fsreturn_ty
+        extern = case fs^.fsabi of
+          RustAbi -> mempty
+          abi -> pretty "extern" <+> viaShow abi <+> mempty
```
If I understand correctly, `viaShow abi` will cause `extern "rust-call" ...` to be pretty-printed as something like `extern RustCall RcNoBody ...`, which doesn't seem desirable. Perhaps we should call `pretty abi` and change the `Pretty Abi` instance to render `RustCall ...` as `"rust-call"`?

spernsteiner replied:
I think the current rendering is fine.  `--print-mir` output is primarily a debugging aid and already only vaguely follows the usual Rust/MIR syntax.  Also, it's important to distinguish the `RcNoSpreadArg` and `RcSpreadArg` in the output.

RyanGlScott commented on crucible-mir/src/Mir/TransTy.hs:
```diff
@@ -314,6 +315,26 @@ isZeroSized col ty = go ty
       _ -> False
 
 
+-- | Get the "ABI-level" function arguments for `sig`, which determines the
+-- arguments we use for the Crucible `FnHandle`.  This applies a special
+-- adjustment for `extern "rust-call"` functions, like the one rustc applies
+-- when converting a `FnSig` to `FnAbi`.
```
```suggestion
-- | Get the \"ABI-level\" function arguments for @sig@, which determines the
-- arguments we use for the Crucible `FnHandle`.  This applies a special
-- adjustment for @extern \"rust-call\"@ functions, like the one rustc applies
-- when converting a `FnSig` to `FnAbi`.
```

spernsteiner replied:
Fixed in 35f00b44baee28f5649bba773df875c95d3b71be

RyanGlScott commented on crucible-mir/src/Mir/TransTy.hs:
```diff
@@ -314,6 +315,26 @@ isZeroSized col ty = go ty
+      [M.TyTuple tys] -> tys
+      [selfTy, M.TyTuple tys] -> selfTy : tys
+      _ -> error $
+          "unexpected argument list for " ++ show (sig ^. M.fsabi) ++  ": "
+            ++ show (sig ^. M.fsarg_tys)
```
Consider using `show (pretty ...)` instead of `show ...` here to make the results a bit more compact and human-readable.

RyanGlScott commented on crucible-mir/src/Mir/Trans.hs:
```diff
@@ -2621,6 +2647,13 @@ transCollection col halloc = do
     hmap2 <- mkShimHandleMap col halloc
     let hmap = hmap1 <> hmap2
 
+    --forM_ (Map.toList hmap) $ \(name, mh) -> do
+    --  MirHandle _ sig handle <- return mh
+    --  traceM $ "function " ++ show name ++ ":"
+    --  traceM $ "  sig = " ++ show sig
+    --  traceM $ "  handle = " ++ show (FH.handleType handle)
```
Perhaps we should always print this if the `debugLevel` (from `MirGenerator`) is high enough?

spernsteiner replied:
Enabled for `-d4` and higher in a16f6f8a9d6cf2155fde79e6433f0f3408b44ba9

RyanGlScott commented on crucible-mir/src/Mir/Trans.hs:
```diff
@@ -2005,27 +2033,29 @@ genFn (M.Fn fname argvars _ body@(MirBody localvars blocks _)) rettype inputs =
+        case (inputs, vars) of
+            ([], []) -> return ()
+            (input : inputs', var : vars') -> do
                 mvi <- use $ varMap . at (var ^. varname)
                 Some vi <- case mvi of
                     Just x -> return x
                     Nothing -> mirFail $ "no varinfo for arg " ++ show (var ^. varname)
-                Refl <- testEqualityOrFail (R.typeOfAtom input) (varInfoRepr vi) $
+                MirExp inputTpr inputExpr <- return input
```
Consider pattern-matching directly on `input` to avoid this `return input` statement:

```hs
            (MirExp inputTpr inputExpr : inputs', var : vars') -> do
```

spernsteiner replied:
Fixed in 579afd3857198af9fc108870f387483beb58c677

RyanGlScott commented on crucible-mir/src/Mir/Trans.hs:
```diff
@@ -1880,6 +1880,7 @@ initLocals localVars addrTaken = forM_ localVars $ \v -> do
         Just (MirExp tpr' val) -> do
             Refl <- testEqualityOrFail tpr tpr' $
                 "initialValue produced " ++ show tpr' ++ " instead of " ++ show tpr
+                  ++ ", for " ++ show ty
```
Slightly more compact:

```suggestion
                  ++ ", for " ++ show (pretty ty)
```

spernsteiner replied:
Fixed in 579afd3857198af9fc108870f387483beb58c677

RyanGlScott reviewed (commented):
Does this fix https://github.com/GaloisInc/crucible/issues/1222? If so, can one or more of the programs from that issue be added as a `crux-mir` regression test?

samcowger commented on crucible-mir/src/Mir/Trans.hs:
```diff
@@ -1973,14 +1974,41 @@ genFn :: HasCallStack =>
+      -- argument is local `_1`), but we convert to a 0-based index for
+      -- convenience.
+      let splitIndex = spreadArg - 1
+      when (splitIndex > length inputExps) $
```
```suggestion
      when (splitIndex >= length inputExps) $
```
IIUC, the case of `splitIndex == length inputExps` is also erroneous, and it gets caught when computing `tupleFieldTys`. However, I think checking for it here would yield a more helpful error message.

samcowger commented on crucible-mir/src/Mir/TransTy.hs:
```diff
@@ -314,6 +315,26 @@ isZeroSized col ty = go ty
       _ -> False
 
 
+-- | Get the "ABI-level" function arguments for `sig`, which determines the
+-- arguments we use for the Crucible `FnHandle`.  This applies a special
+-- adjustment for `extern "rust-call"` functions, like the one rustc applies
+-- when converting a `FnSig` to `FnAbi`.
```
Perhaps state that the "special adjustment" is untupling?

spernsteiner replied:
I adjusted this comment slightly in 35f00b44baee28f5649bba773df875c95d3b71be.  I don't think it's necessary to be specific about the adjustment - the point is that this function "does the right thing" in all cases, including `extern "rust-call"`.

samcowger commented on crucible-mir/src/Mir/Trans.hs:
```diff
@@ -1973,14 +1974,41 @@ genFn :: HasCallStack =>
+      -- `inputExps` to match `argvars` by tupling up some of the arguments.
+      --
+      -- Arguments from `spreadArg` onward need to be tupled.  Note that
+      -- `fsspreadarg` is 1-based (it's a MIR `Local` ID, and the first
```
Do you mean `spreadArg` here, instead of `fsspreadarg`?

spernsteiner replied:
Fixed in 35f00b44baee28f5649bba773df875c95d3b71be

RyanGlScott commented:
To answer my question above: yes, this PR does appear to fix #1222, as I can now run the program from https://github.com/GaloisInc/crucible/issues/1222#issuecomment-2233313541 without issue. Moreover, the `crux-mir/test/conc_eval/clos/fn_dyn.rs` test case is very nearly identical to that program, so we have test coverage.

Still, could you amend the PR or commit message to say "Fixes #1222" somewhere so that this connection is more obvious? It might also be worth mentioning this change in the `crucible-mir` changelog.

spernsteiner commented:
I added some more test cases that cover additional `dyn Fn` variants.

It turns out there are still issues with `Box<dyn FnOnce>` because the trait definition for `FnOnce::call_once` has a signature of `fn(dyn FnOnce, ...)`, which triggers the "standalone use of `dyn`" error.  This would require some additional work, since `dyn Trait` methods with by-value `self` aren't normally supported (if you try to write your own version of `FnOnce`, you won't be able to call it through `Box<dyn MyFnOnce>` like you can with `Box<dyn FnOnce>`).  I've still marked the PR as fixing #1222 because the `FnOnce` case results in a different error message.

There's a separate issue with `Box<dyn Fn>`/`Box<dyn FnMut>`: when the box is dropped, symbolic execution enters infinite recursion in `drop_in_place`.  I think this will likely be fixed by #1445.

I also updated the changelog to mention that calling `dyn Fn` and `dyn FnMut` (but not `dyn FnOnce`) is now supported.
